appletree/plot.py

```diff
@@ -0,0 +1,302 @@
+                if prior["prior_type"] == "norm":
+                    ax.plot(x, self._norm_pdf(x, **prior["prior_args"]), color="b")
+                elif prior["prior_type"] == "uniform":
+                    ax.plot(x, self._uniform_pdf(x, **prior["prior_args"]), color="b")
+                elif prior["prior_type"] == "twohalfnorm":
+                    ax.plot(x, self._thn_pdf(x, **prior["prior_args"]), color="b")
+
+        return fig, axes
+
+    def plot_autocorr(self, fig=None, **plot_kwargs):
+        """Plot the autocorrelation time of each parameter, as the diagnostic of the convergence.
+
+        Args:
+            fig: the figure to plot on. If None, a new figure will be created.
+            plot_kwargs: the keyword arguments passed to plt.plot().
+        Returns:
+            fig: the figure.
+            axes: the axes of the figure.
+
+        """
+        n_cols = 2
+        n_rows = int(np.ceil(self.n_param / n_cols))
+
+        if fig is None:
+            fig = plt.figure(figsize=(10, 3 * n_rows))
+        plot_kwargs.setdefault("marker", "o")
+
+        def autocorr_func_1d(x, norm=True):
+            x = np.atleast_1d(x)
+            if len(x.shape) != 1:
+                raise ValueError("invalid dimensions for 1D autocorrelation function")
+            n = next_pow_two(len(x))
+            f = np.fft.fft(x - np.mean(x), n=2 * n)
+            acf = np.fft.ifft(f * np.conjugate(f))[: len(x)].real
+            acf /= 4 * n
+            if norm:
+                acf /= acf[0]
+            return acf
+
+        def next_pow_two(n):
+            i = 1
+            while i < n:
+                i = i << 1
+            return i
+
+        def auto_window(taus, c):
+            m = np.arange(len(taus)) < c * taus
+            if np.any(m):
+                return np.argmin(m)
+            return len(taus) - 1
+
+        def autocorr_new(y, c=5.0):
+            f = np.zeros(y.shape[1])
+            for yy in y:
+                f += autocorr_func_1d(yy)
+            f /= len(y)
+            taus = 2.0 * np.cumsum(f) - 1.0
+            window = auto_window(taus, c)
+            return taus[window]
+
+        assert self.n_iter > 1000, "The chain is too short to compute the autocorrelation time"
+        N = np.geomspace(100, self.n_iter, 10).astype(int)
+        axes = []
+        for i in range(self.n_param):
+            chain = self.chain[:, :, i].T
+            tau = np.empty(len(N))
+            for j, n in enumerate(N):
+                tau[j] = autocorr_new(chain[:, :n])
+
+            ax = fig.add_subplot(n_rows, n_cols, i + 1)
+            ax.plot(N, tau, label="Sample estimation", **plot_kwargs)
+            ax.plot(N, N / 50, "k--", label="N / 50")
+            ax.set_xscale("log")
+            ax.set_yscale("log")
+            ax.set_xlabel("Number of iterations")
+            ax.set_ylabel(f"Auto correlation of {self.param_names[i]}")
+            ax.legend()
```

Comment:
You can set the legend for the whole plot, but not for individual subplot. Also here maybe you can just set the xlabel of the bottom two subplots.

Reply:
Like
```
        handles, labels = ax.get_legend_handles_labels()
        fig.legend(
            loc='lower center',
            handles=handles, labels=labels,
            bbox_to_anchor=(0.5, 1.0),
        )
```